tchannel/tchannel.py

```diff
@@ -315,14 +322,36 @@ def advertise(self, routers=None, name=None, timeout=None,
                 log.exception('Failed to read seed routers list.')
                 raise
 
-        dep_result = yield self._dep_tchannel.advertise(
-            routers=routers,
-            name=name,
-            timeout=timeout
-        )
-
-        body = yield dep_result.get_body()
-        headers = yield dep_result.get_header()
-        response = Response(json.loads(body), headers or {})
-
-        raise gen.Return(response)
+        @gen.coroutine
+        def _advertise():
+            result = yield self._dep_tchannel.advertise(
+                routers=routers,
+                name=name,
+                timeout=timeout,
+            )
+            body = yield result.get_body()
+            headers = yield result.get_header()
+            response = Response(json.loads(body), headers or {})
+            raise gen.Return(response)
+
+        def on_advertise(future):
+            if not future.exception():
+                return
+
+            # If the request failed, clear the response so that we can try
+            # again.
+            with self._advertise_lock:
+                # `is` comparison to ensure we're not deleting another Future.
+                if self._advertise_response is future:
+                    self._advertise_response = None
+
+        is_new = False
+        with self._advertise_lock:
+            if self._advertise_response is None:
+                self._advertise_response = _advertise()
+                is_new = True
+            future = self._advertise_response
+
+        if is_new:
```

Comment:
No it won't. That's if you use `add_future`. If you use `add_done_callback`, it will call the callback right away as part of the `add_done_callback` call. See http://tornado.readthedocs.org/en/latest/_modules/tornado/concurrent.html#Future.add_done_callback.